### PR TITLE
Fix bond checkpointing

### LINKED ISSUE #2040: `system.part.bonds` cannot be pickled for Cython version < 0.26.0
Pickle `system.part[:].bonds` fails.

Example script:
```python
import espressomd
from espressomd.interactions import HarmonicBond
import pickle
 
system = espressomd.System(box_l=[5., 5., 5.])
harm = HarmonicBond(r_0=1., k=200.)
system.bonded_inter.add(harm)

system.part.add(pos=[1., 1., 1.], id=0, type=0)
system.part.add(pos=[2., 1., 1.], id=1, type=0)

system.part[0].add_bond((harm, 1))

pickle.dump({"bonds": system.part[:].bonds}, open("particle_bonds.dat", "w")) 
```
fails with
```
Exception: The constructor has to be called either with a bond id (as interger), or with a set of keyword arguments describing a new interaction
```

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Fixes #2040 
Description of changes:
 - unpickle bonded_interactions before particles
 - added test for bonded interaction pickling
 - using vector for bonded interaction parameters in the core

PR Checklist
------------
 - [x] Tests?
   - [x] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

File: src/python/espressomd/interactions.pxd [hunk 1]
```diff
@@ -463,14 +463,13 @@ cdef extern from "interaction_data.hpp":
         IBM_Tribend_Parameters ibm_tribend
         IBM_VolCons_Parameters ibmVolConsParameters
 
-    ctypedef struct bonded_ia_parameters:
+    cdef struct Bonded_ia_parameters:
```
Comment (pkreissl) [inline src/python/espressomd/interactions.pxd:488]: Shouldn't this then be `Bonded_IA_parameters`, for consistencie's sake? I get that you just want to capitalize the first letter, however I am not entirely happy with the multitude of `B/bond(ed)*ia/IA*param(eters)` variables. Not saying that it was much better before...

**Review by pkreissl — COMMENTED** [review]
Compilation fails for me with
```
/path_to_espresso/build/src/python/espressomd/analyze.cpp:1512:8: error: ‘Harmonic_dumbbell_bond_parameters’ does not name a type
 static Harmonic_dumbbell_bond_parameters __pyx_convert__from_py_Harmonic_dumbbell_bond_parameters(PyObject *); /*proto*/
        ^
/path_to_espresso/build/src/python/espressomd/analyze.cpp:20927:8: error: ‘Harmonic_dumbbell_bond_parameters’ does not name a type
 static Harmonic_dumbbell_bond_parameters __pyx_convert__from_py_Harmonic_dumbbell_bond_parameters(PyObject *__pyx_v_obj) {
        ^
```
~using `maxset.hpp`~  not specifying `myconfig.hpp`.

**Comment by pkreissl** [comment]
Doesn't fix  issue #2040 for me. Error stays the same using Cython version 0.23.4 (std in Ubuntu 16.04.4 LTS)

File: src/python/espressomd/interactions.pxd (hunk 1 above)
Comment (KaiSzuttor) [reply src/python/espressomd/interactions.pxd:488]: i have not renamed it, i just used the core name here

**Comment by KaiSzuttor** [comment]
in the issue, you said for version <0.26, i tested against version 0.25

**Comment by pkreissl** [comment]
I just rebuilt making sure I am actually using Cython 0.25. Issue stays the same for me.

**Comment by KaiSzuttor** [comment]
okay, maybe you should dig into this further, at least we now have tested checkpointing for bonded interactions which we should have before the release

**Comment by KaiSzuttor** [comment]
i added a ```__reduce__``` function, which actually made the checkpointing of bonds working also without specifying the pickle protocol

**Comment by pkreissl** [comment]
Test `#67: coulomb_mixed_periodicity` consistently fails on my machine (impala) with:
```
.ERROR: number of cells 3 is smaller than minimum 4 (interaction range too large or min_num_cells too large)
ERROR: number of cells 3 is smaller than minimum 4 (interaction range too large or min_num_cells too large)
ERROR: number of cells 3 is smaller than minimum 4 (interaction range too large or min_num_cells too large)
ERROR: number of cells 3 is smaller than minimum 4 (interaction range too large or min_num_cells too large)
E
======================================================================
ERROR: test_zz_p3mElc (__main__.CoulombMixedPeriodicity)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/pathtoespresso/build/testsuite/coulomb_mixed_periodicity.py", line 118, in test_zz_p3mElc
    self.S.actors.add(p3m)
  File "espressomd/actors.pyx", line 186, in espressomd.actors.Actors.add (/pathtoespresso/build/src/python/espressomd/actors.cpp:5237)
  File "espressomd/actors.pyx", line 49, in espressomd.actors.Actor._activate (/pathtoespresso/build/src/python/espressomd/actors.cpp:2009)
  File "espressomd/electrostatics.pyx", line 347, in espressomd.electrostatics.P3M._activate_method (/pathtoespresso/build/src/python/espressomd/electrostatics.cpp:5138)
  File "espressomd/electrostatics.pyx", line 338, in espressomd.electrostatics.P3M._tune (/pathtoespresso/build/src/python/espressomd/electrostatics.cpp:4910)
  File "espressomd/electrostatics.pxd", line 134, in espressomd.electrostatics.python_p3m_adaptive_tune (/pathtoespresso/build/src/python/espressomd/electrostatics.cpp:11013)
  File "espressomd/utils.pyx", line 268, in espressomd.utils.handle_errors (/pathtoespresso/build/src/python/espressomd/utils.cpp:7111)
Exception: Error in p3m_adaptive_tune
```
using the std Ubuntu 16.04 Cython version 0.23.4 and `maxset.hpp`.

**Comment by KaiSzuttor** [comment]
whats the difference to the docker?

**Comment by pkreissl** [comment]
This seems to be related to the number of cores used while testing.
If I run the test manually on X cores:
`mpirun -np X ./pypresso ./testsuite/coulomb_mixed_periodicity.py`
the test passes up to `X=3` and fails for `X>=4` (checked up to 7).
How many cores does the docker testing run on?

**Comment by KaiSzuttor** [comment]
The CI only tests with 2/3 processes.

**Comment by KaiSzuttor** [comment]
actually the current python branch fails on 4 cores too. So I don't think that this PR is related to the failing test.

**Comment by pkreissl** [comment]
Just realized that, too.
I ~am opening~ opened ~an~ issue #2061.

**Review by pkreissl — APPROVED** [review]
Looks good to me.

**Pull request merged**